protocol/src/protocol.calls.proposed.md

```diff
@@ -0,0 +1,110 @@
+
+#### Calls
+
+The LSP provides retrieving the call hierachy information with the following request.
+
+_Client Capabilities_:
+
+```ts
+CallsClientCapabilities {
+    /**
+     * The text document client capabilities
+     */
+    textDocument?: {
+        /**
+         * Capabilities specific to the `textDocument/calls`
+         */
+        calls?: {
+            /**
+             * Whether implementation supports dynamic registration. If this is set to `true`
+             * the client supports the new `(TextDocumentRegistrationOptions & StaticRegistrationOptions)`
+             * return value for the corresponding server capability as well.
+             */
+            dynamicRegistration?: boolean;
+        };
+}
+```
+
+_Server Capabilities_:
+
+```ts
+CallsServerCapabilities {
+    /**
+     * The server provides Call Hierarchy support.
+     */
+    callsProvider?: boolean | (TextDocumentRegistrationOptions & StaticRegistrationOptions);
+}
+```
+
+##### Calls Request
+
+_Request_:
+
+The `textDocument/calls` request is sent from the client to the server to resolve the callers or callees of a symbol at a given text document position. Returns the callers or callees corresponding to the requested direction.
+
+* method: ‘textDocument/calls’
+* params: `CallsParams` defined as follows:
+
+```ts
+export interface CallsParams extends TextDocumentPositionParams {
+    /**
+     * Outgoing direction for callees.
+     * The default is incoming for callers.
+     */
+    direction?: CallDirection;
+}
+
+export enum CallDirection {
+    /**
+     * Incoming calls aka. callers
+     */
+    Incoming = "incoming",
+    /**
+     * Outgoing calls aka. callees
+     */
+    Outgoing = "outgoing",
+}
+```
+
+_Response_:
+
+The server will send a `CallsResult` object containing a `symbol` of a definition which it finds for given text document position. The symbol will be undefined, if no such definition is found.
+
+The result object contains a list of `Call`s which include more information about the call site. It also contains a symbol location of the caller or collee which can be used by clients to make a request for the next level of the call hierarchy.
+
+* result: `CallsResult` defined as follows:
+
+```ts
+export interface CallsResult {
+    /**
+     * The symbol of a definition for which the request was made.
+     *
+     * If no definition is found at a given text document position, the symbol is undefined.
+     */
+    symbol?: DocumentSymbol;
+    /**
+     * List of calls.
+     */
+    calls: Call[];
+}
+
+export interface Call {
+    /**
+     * Direction of the call.
+     */
+    direction: CallDirection;
```

Comment:
I think this field is implied by the call it is a result of, no?

Reply:
That's correct.